spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseExpireSnapshotsSparkAction.java

```diff
@@ -212,11 +214,21 @@ private String jobDesc() {
   }
 
   private ExpireSnapshots.Result doExecute() {
-    boolean streamResults = PropertyUtil.propertyAsBoolean(options(), STREAM_RESULTS, false);
-    if (streamResults) {
-      return deleteFiles(expire().toLocalIterator());
+    boolean parallelDelete = PropertyUtil.propertyAsBoolean(options(), PARALLEL_DELETE, false);
+
+    if (parallelDelete) {
+      Table serializableTable = SerializableTable.copyOf(table);
+
+      return expire().javaRDD().mapPartitions((Iterator<Row> iter) ->
+        Arrays.asList(
+            ActionUtils.deleteFiles(iter, (file) -> serializableTable.io().deleteFile(file), null)
+        ).iterator()
+      ).reduce((r1, r2) -> new BaseExpireSnapshotsActionResult(
+          r1.deletedDataFilesCount() + r2.deletedDataFilesCount(),
+          r1.deletedManifestsCount() + r2.deletedManifestsCount(),
+          r1.deletedManifestListsCount() + r2.deletedManifestListsCount()));
```

Comment:
Using Spark to delete files can easily result in too high of a request rate, which causes S3 to throttle and send 503, "Slow down", responses. When that happens, files are probably leaked as orphans. That's why we have a way to parallelize locally, but not to use Spark.

This action also exposes the `expire()` method that returns a `Dataset` you can use if you want to run the expiration like this in your environment.